transport/http2_client.go

```diff
@@ -289,7 +289,10 @@ func (t *http2Client) NewStream(ctx context.Context, callHdr *CallHdr) (_ *Strea
 		}
 	}
 	if _, err := wait(ctx, t.shutdownChan, t.writableChan); err != nil {
-		// t.streamsQuota will be updated when t.CloseStream is invoked.
+		// Return the quota back now because there is no stream returned to the caller.
+		if _, ok := err.(StreamError); ok && checkStreamsQuota {
```

Comment:
Why does the type of error returned from the wait method matter here at all?  The wait method can also return ErrConnClosing if the shutdownChan is readable but it is not of type StreamError.

I believe that per the original description of the problem, involving the pseudo random behavior of the Go select statement, it could have been readable on the invocation on line 282 and you get unlucky and quota gets picked instead.  You could argue that in a shutdown event you will tear everything down so the inaccuracy of the quota is irrelevant, but it is probably best to have the least number of special cases.  To that end, if you are here the type of error shouldn't matter.

Perhaps I missed something?